// postcss-cssnext/lib/features.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = {
  // Reminder: order is important
  customProperties: function customProperties(options) {
    return require("postcss-custom-properties")(options);
  },
  calc: function calc(options) {
    return require("postcss-calc")(options);
  },
  nesting: function nesting(options) {
    return require("postcss-nesting")(options);
  },
  customMedia: function customMedia(options) {
    return require("postcss-custom-media")(options);
  },
  mediaQueriesRange: function mediaQueriesRange(options) {
    return require("postcss-media-minmax")(options);
  },
  customSelectors: function customSelectors(options) {
    return require("postcss-custom-selectors")(options);
  },
  colorRebeccapurple: function colorRebeccapurple(options) {
    return require("postcss-color-rebeccapurple")(options);
  },
  colorHwb: function colorHwb(options) {
    return require("postcss-color-hwb")(options);
  },
  colorGray: function colorGray(options) {
    return require("postcss-color-gray")(options);
  },
  colorHexAlpha: function colorHexAlpha(options) {
    return require("postcss-color-hex-alpha")(options);
  },
  colorFunction: function colorFunction(options) {
    return require("postcss-color-function")(options);
  },
  fontVariant: function fontVariant(options) {
    return require("postcss-font-variant")(options);
  },
  filter: function filter(options) {
    return require("pleeease-filters")(options);
  },
  initial: function initial(options) {
    return require("postcss-initial")(options);
  },
  rem: function rem(options) {
    return require("pixrem")(options);
  },
  pseudoElements: function pseudoElements(options) {
    return require("postcss-pseudoelements")(options);
  },
  pseudoClassMatches: function pseudoClassMatches(options) {
    return require("postcss-selector-matches")(options);
  },
  pseudoClassNot: function pseudoClassNot(options) {
    return require("postcss-selector-not")(options);
  },
  pseudoClassAnyLink: function pseudoClassAnyLink(options) {
    return require("postcss-pseudo-class-any-link")(options);
  },
  colorRgba: function colorRgba(options) {
    return require("postcss-color-rgba-fallback")(options);
  },
  autoprefixer: function autoprefixer(options) {
    return require("autoprefixer")(options);
  }
};
module.exports = exports["default"];
